### PR TITLE
Issue 79 undefined in section subsection

### PR DESCRIPTION
**Pull request by KamBustos**
The original bug was that `undefined` appears in section and subsection on ng.html when parsing. I fixed the bug so that it no longer displays `undefined` as a result. I edited the ng.js file specifically parseSections(). You can test this fix by typing Kam 5-7 when adding a new Canonical Reference on my most recent commit. 

Bug:
issue79_bug
Fixed:
issue79_fixed

`undefined` was also appearing when loading existing data. The string "undefined" should never be exposed to the user. This was also resolved. You can test this fix by typing Kameron 5:undefined

Bug:
79issueBugUndefined
Fixed:
79issueUndefinedFix

### CODE REVIEW

**Review by thehabes — CHANGES_REQUESTED** [review]
Hi Kam,

Unfortunately these changes were not successful.  Though I cannot cause undefined when I type in new values, there are undefined on page load.

To continue coding, first stage your code for dev.  You did not update your properties.json

```
{
    "ngCollection" : "https://devstore.rerum.io/v1/id/610c54deffce846a83e70625",
    "msCollection" : "https://devstore.rerum.io/v1/id/610ad6f1ffce846a83e70613",
    "generator" : "http://devstore.rerum.io/v1/id/5afeebf3e4b0b0d588705d90",
    "tiny" : "https://tinydev.rerum.io/app",
    "rerum" : "https://devstore.rerum.io/v1"
}
```

Once you make this update, load up http://localhost:8080/ng.html#https://devstore.rerum.io/v1/id/614cf6d8e74876243131a67a.  You will see there are some 'undefined' values.

image

**Review by thehabes — CHANGES_REQUESTED** [review]
Upon testing, I am still seeing an undefined.

image

As stated before, make sure to load up http://localhost:8080/ng.html#https://devstore.rerum.io/v1/id/614cf6d8e74876243131a67a and check for instances of undefined.

**Comment by thehabes** [comment]
`undefined` appearing in Canonical Reference Locator is being moved to its own issue and scope.  I will re-review and ignore that for now.

File: js/deer-config.js [hunk 1]
```diff
@@ -689,4 +689,4 @@ function debounce(func,timeout = 500) {
         clearTimeout(timeRemains)
         timeRemains = setTimeout(()=>func.apply(this,args),timeRemains)
     }
-}
```
Comment (thehabes) [inline js/deer-config.js:692]: Oops!  Not a big deal.  If you can, undo this change so there is no diff on this file.

File: js/ng.js [hunk 2]
```diff
@@ -162,17 +162,39 @@ addEventListener('deer-updated', event => {
...
+    }
+
+    // Split the Canonical Reference Locator value using a regex pattern
+    const canonSplit = canonValue.split(/[\s\:\.,;\|#§]/);
+
+    // Iterate through the input fields and populate them with corresponding parts of the split value
+    elemSet.forEach((el, index) => {
+        if (index < canonSplit.length) {
+            // Check if the split part is not "undefined" before assignment
+            if (canonSplit[index] !== "undefined") {
```
Comment (thehabes) [inline js/ng.js:186]: "undefined" may be a string, but it may also just be the javascript primitive `undefined` so make sure to check that as well.

File: js/ng.js [hunk 3]
```diff
@@ -162,17 +162,39 @@ addEventListener('deer-updated', event => {
...
+                el.value = canonSplit[index];
+            } else {
+                el.value = ''; // Set to an empty string if the split part is "undefined"
```
Comment (cubap) [inline js/ng.js:189]: When you find yourself doing the same thing twice, consider inverting the condition and just leaving the block early instead.

File: js/ng.js [hunk 4]
```diff
@@ -162,17 +162,39 @@ addEventListener('deer-updated', event => {
...
 
 function parseSections() {
-    const canonValue = document.querySelector('input[deer-key="canonicalReference"]')?.value
-    const _document = document.querySelector('input[deer-key="_document"]')
-    const _section = document.querySelector('input[deer-key="_section"]')
-    const _subsection = document.querySelector('input[deer-key="_subsection"]')
-    const elemSet = [_document, _section, _subsection]
-    if (elemSet.includes(null) || !canonValue?.length) { throw new Error(`Missing elements in ${elemSet.join(', ')}`) }
+    // Get the Canonical Reference Locator value
+    const canonValue = document.querySelector('input[deer-key="canonicalReference"]')?.value;
```
Comment (cubap) [inline js/ng.js:166]: The style of the rest of the code doesn't use ";"

File: js/ng.js (hunk 2 above)
Comment (cubap) [inline js/ng.js:186]: Preventing the `undefined` above may be more robust than checking for its introduction here.

File: js/ng.js [hunk 5]
```diff
@@ -162,17 +162,39 @@ addEventListener('deer-updated', event => {
...
+    // Check if any of the input fields are missing or if the Canonical Reference Locator is empty
+    if (elemSet.includes(null) || !canonValue?.length) {
+        throw new Error(`Missing elements in ${elemSet.join(', ')}`);
+    }
+
+    // Split the Canonical Reference Locator value using a regex pattern
+    const canonSplit = canonValue.split(/[\s\:\.,;\|#§]/);
+
+    // Iterate through the input fields and populate them with corresponding parts of the split value
+    elemSet.forEach((el, index) => {
```
Comment (cubap) [inline js/ng.js:183]: a `while` loop here or looping over the split would prevent having to do any checking, though this "works"

**Review by cubap — COMMENTED** [review]
Good look. I think there are just a few touches I would recommend.

**Review by thehabes — CHANGES_REQUESTED** [review]
Good work here, you have stopped undefined from appearing when clicking "Parse Sections".  I just have a couple notes for clean up and future proofing.

Also, please get rid of the unnecessary semicolons (the ones ending a line of code).

File: js/ng.js (hunk 5 above)
Comment (thehabes) [reply js/ng.js:183]: Agreed, this has a way to be a little simpler and perhaps a little cleaner.  However, it works and I am OK with it moving on up.

**Review by thehabes — APPROVED** [review]
That does it for me if @AbhilashKotha approves.

File: js/ng.js (hunk 5 above)
Comment (KamBustos) [reply js/ng.js:183]: > a `while` loop here or looping over the split would prevent having to do any checking, though this "works"

I can optimize this in the new issue I created (Issue 96). I'll make a note of it.

**Review by AbhilashKotha — APPROVED** [review]
lgtm
